src/app/shared/components/course-form/course-form.component.ts

```diff
@@ -1,19 +1,105 @@
+import { v4 as uuidv4 } from 'uuid';
+
 
 @Component({
   selector: 'app-course-form',
   templateUrl: './course-form.component.html',
   styleUrls: ['./course-form.component.scss'],
 })
-export class CourseFormComponent {
+export class CourseFormComponent{
+  removeIconName:IconNames = IconNames.TrashCan;
+  removeIcon: IconProp | undefined = ['fas',this.removeIconName as IconName];
+  courseForm!:FormGroup;
+  allAuthors:Author[]=[];
+/*   courseAuthors: Author[]=[]; */
+  nonCourseAuthors: Author[]=[];
+  submitted:boolean = false;
+  wrongCreation:boolean = false;
+
   constructor(public fb: FormBuilder, public library: FaIconLibrary) {
     library.addIconPacks(fas);
+    this.buildForm();
+  }
+
+  buildForm(): void {
+    this.courseForm = this.fb.group({
+      title: ['',[Validators.required,Validators.minLength(2)]],
+      description: ['',[Validators.required,Validators.minLength(2)]],
+      newAuthor: this.fb.group({
+        author: ['',[Validators.required, Validators.minLength(2), this.authorValidator()]]
+      }),
+      duration: ['',[Validators.required, Validators.min(0)]],
+      authors: this.fb.array([])  // FormArray
+    });
+  }
+
+  addAuthorToCourse(author:Author):void {
+    this.authors.push(this.fb.control(author));
+    this.nonCourseAuthors = this.nonCourseAuthors.filter((nonCourseAuthor)=>nonCourseAuthor.id != author.id); 
+  }
+  removeAuthorFromCourse(author:Author, arrayIndex:number):void {
+    this.nonCourseAuthors = [...this.nonCourseAuthors, author];
+    this.authors.removeAt(arrayIndex);
   }
-  courseForm!: FormGroup;
   // Use the names `title`, `description`, `author`, 'authors' (for authors list), `duration` for the form controls.
+  createCourse():void {
+  }
+  cancelCourse():void {
+  }
+  addAuthor():void {
+    if(this.author?.valid && 
+       this.author?.value )
+    {
+      const newAuthor:Author = {id:this.generateId(), name:this.author!.value};
+      this.allAuthors.push(newAuthor);
+      this.nonCourseAuthors.push(newAuthor);
+      this.courseForm.get('author')?.reset();
+      this.wrongCreation=false;
+    }
+    else this.wrongCreation=true;
+  }
+  private generateId(): string {
+    return uuidv4();
+  }
+  get title():AbstractControl | null {
+    return this.courseForm.get('title');
+  }
+  get description():AbstractControl | null  {
+    return this.courseForm.get('description');
+  }
+  get duration():AbstractControl | null  {
+    return this.courseForm.get('duration');
+  }
+  get author():AbstractControl | null  {
+    return this.courseForm.get('newAuthor.author');
+  }
+  get authors(): FormArray {
+    return this.courseForm.get('authors') as FormArray;
+  }
+
+  authorValidator(): ValidatorFn {
+    return (control: AbstractControl): ValidationErrors | null => {
+      const valid = /^[a-zA-Z0-9\s]+$/.test(control.value);
+      return !valid ? {invalidAuthorCharacter: 'Author name can have only latin characters or numbers'} : null;
```

Comment:
You sure this error msg is displayed on the templet once the validator returns? I can see the same msg declared on the template.
![image](https://github.com/user-attachments/assets/ecc957b4-1ea4-4c5f-bce6-5b580f89d97d)

Reply:
Nice , thank you! I forgot to change the fix content for the validators error content